File: src/main/java/com/ledgerguard/detection/ml/Agreement.java
package com.ledgerguard.detection.ml;

/**
 * How the statistical composite and the isolation score relate for one account.
 *
 * <h2>Why the two scores are never blended</h2>
 *
 * Folding the model in as a sixth Phase 8 signal was considered and rejected,
 * for four reasons that all point the same way:
 *
 * <ol>
 *   <li>It would need a weight — a sixth unfitted judgement stacked on the five
 *       Phase 8 already carries.</li>
 *   <li>It would make the composite <em>partly unexplainable</em>. Every
 *       statistical signal can state its reasoning in a sentence; a forest score
 *       cannot. Blending them produces a number where part of the provenance
 *       simply stops.</li>
 *   <li><b>Disagreement is the most useful thing here.</b> An account the
 *       statistics call quiet and the model calls extreme is the single most
 *       interesting row in the system: either the model has found structure the
 *       signals cannot express, or the model is wrong, and both are worth
 *       knowing. Averaging destroys exactly that information.</li>
 *   <li>Neither score is validated. Combining two unvalidated numbers produces
 *       one unvalidated number that <em>looks more authoritative for being
 *       single</em>, which is the specific failure this phase is meant to avoid.</li>
 * </ol>
 *
 * <p>So both scores travel side by side, and this enum names their relationship
 * rather than resolving it.
 *
 * <h2>The thresholds</h2>
 *
 * "Elevated" needs a cut-off on each scale, and both are <b>conventions, not
 * fitted values</b> — the same caveat that applies to Phase 8's weights.
 *
 * <p>They are deliberately different numbers. The statistical composite is a
 * weighted mean of scores that are zero until their signal fires, so 0.5 there
 * means a substantial share of the signals fired. An isolation score is
 * concentrated around 0.5 <em>by construction</em> — that is the score of a point
 * at average depth — so the same cut-off would call half the population
 * elevated. Using one number for both scales would be tidier and wrong.
 */
public enum Agreement {

    /** Neither score is elevated. */
    BOTH_QUIET,

    /** Both are. The clearest case, and the only one where the two corroborate. */
    BOTH_ELEVATED,

    /**
     * The statistics fired and the model did not. Often means the behaviour is
     * unusual for this account but ordinary across the population.
     */
    STATISTICAL_ONLY,

    /**
     * The model fired and the statistics did not. The row worth reading first:
     * either a combination of features no single signal is shaped to catch, or
     * the model reacting to something that does not matter.
     */
    ML_ONLY;

    /** A composite at or above this counts as elevated. Convention, not fitted. */
    public static final double STATISTICAL_ELEVATED = 0.5;

    /**
     * An isolation score at or above this counts as elevated. Convention, not
     * fitted, and higher than it looks: see the class note on why 0.5 would be
     * meaningless on this scale.
     */
    public static final double ML_ELEVATED = 0.6;

    public static Agreement of(double statisticalScore, double mlScore) {
        boolean statistical = statisticalScore >= STATISTICAL_ELEVATED;
        boolean ml = mlScore >= ML_ELEVATED;

        if (statistical && ml) {
            return BOTH_ELEVATED;
        }
        if (statistical) {
            return STATISTICAL_ONLY;
        }
        if (ml) {
            return ML_ONLY;
        }
        return BOTH_QUIET;
    }
}

File: console/src/api/agreement.ts
import type { AgreementState } from './types';

/**
 * The agreement state as the console renders it, which is finer than the enum.
 *
 * `Agreement` has four values. Two of them cover pairs of situations that mean
 * materially different things, and the backend already publishes enough to tell
 * them apart — so collapsing them here would be the console throwing away
 * information the API went to the trouble of exposing.
 *
 * <ul>
 *   <li><b>BOTH_ELEVATED</b> splits on `corroborated`. `Reconciliation.java`
 *       calls the false case the hardest one in the system: two elevated scores
 *       resting on unrelated evidence, which looks like corroboration in the
 *       numbers and is not.</li>
 *   <li><b>ML_ONLY</b> splits three ways. If a signal did fire but the composite
 *       stayed under the line, the layers are not disagreeing at all — that is
 *       dilution, and calling it a disagreement would misreport it. Otherwise it
 *       matters whether the model isolated on axes the statistical layer cannot
 *       see, or on axes it shares and read differently.</li>
 * </ul>
 *
 * Nothing here is inferred beyond what the fields say. Each branch is a
 * restatement of a condition `Reconciliation.narrate` already switches on.
 */
export type DisplayState =
  | 'NO_MODEL'
  | 'BOTH_QUIET'
  | 'STATISTICAL_ONLY'
  | 'BOTH_ELEVATED_SHARED_AXIS'
  | 'BOTH_ELEVATED_UNRELATED'
  | 'ML_ONLY_DILUTED'
  | 'ML_ONLY_OUTSIDE_VIEW'
  | 'ML_ONLY_SAME_AXIS';

export interface DisplayAgreement {
  state: DisplayState;
  /** The enum value, always shown, so the console's finer reading never hides the API's. */
  base: AgreementState | null;
  /** Short heading. Never the word "flagged": the whole point is that the states differ. */
  label: string;
  /** The distinction the sub-state exists to draw. Empty where there is no sub-state. */
  qualifier: string;
  /** A sentence a reviewer can act on, in the API's own terms. */
  reading: string;
  statisticalElevated: boolean;
  modelElevated: boolean;
  /** Category hue token. Encodes *which* state, never *how bad*. */
  tone: 'neutral' | 'statistical' | 'model' | 'both';
}

export interface AgreementInputs {
  agreement: AgreementState | null;
  corroborated: boolean;
  modelDriversOutsideView: string[];
}

export function displayAgreement(input: AgreementInputs): DisplayAgreement {
  const { agreement, corroborated, modelDriversOutsideView } = input;

  if (agreement === null) {
    return {
      state: 'NO_MODEL',
      base: null,
      label: 'No model trained',
      qualifier: 'One layer only',
      reading:
        'There is no isolation score to compare the composite against. This is an absence of a ' +
        'second opinion, not a second opinion that found nothing.',
      statisticalElevated: false,
      modelElevated: false,
      tone: 'neutral',
    };
  }

  switch (agreement) {
    case 'BOTH_QUIET':
      return {
        state: 'BOTH_QUIET',
        base: agreement,
        label: 'Neither layer elevated',
        qualifier: '',
        reading:
          'Quiet in both layers is the weakest claim either can make, not a clean bill of health.',
        statisticalElevated: false,
        modelElevated: false,
        tone: 'neutral',
      };

    case 'STATISTICAL_ONLY':
      return {
        state: 'STATISTICAL_ONLY',
        base: agreement,
        label: 'Statistical layer only',
        qualifier: '',
        reading:
          'Unusual against this account’s own history and ordinary across the training ' +
          'population is the ordinary reading. The other reading is that the model’s features ' +
          'do not capture what the signals caught.',
        statisticalElevated: true,
        modelElevated: false,
        tone: 'statistical',
      };

    case 'BOTH_ELEVATED':
      return corroborated
        ? {
            state: 'BOTH_ELEVATED_SHARED_AXIS',
            base: agreement,
            label: 'Both layers elevated',
            qualifier: 'on a shared axis',
            reading:
              'The two layers point at the same behaviour. This is the only real corroboration ' +
              'available here, and it may still cover only part of the model’s reasoning.',
            statisticalElevated: true,
            modelElevated: true,
            tone: 'both',
          }
        : {
            state: 'BOTH_ELEVATED_UNRELATED',
            base: agreement,
            label: 'Both layers elevated',
            qualifier: 'on unrelated evidence',
            reading:
              'Agreeing that something is unusual is not the same as agreeing on what. No ' +
              'statistical signal shares an axis with anything the model isolated on, so nothing ' +
              'here corroborates anything.',
            statisticalElevated: true,
            modelElevated: true,
            tone: 'both',
          };

    case 'ML_ONLY':
      if (corroborated) {
        return {
          state: 'ML_ONLY_DILUTED',
          base: agreement,
          label: 'Model only',
          qualifier: 'but a signal did fire — diluted, not disagreeing',
          reading:
            'A statistical signal fired on an axis the model also isolated on, and the composite ' +
            'still stayed under the 0.50 convention. On that axis the layers are not disagreeing; ' +
            'the statistical score simply did not carry it over the line.',
          statisticalElevated: false,
          modelElevated: true,
          tone: 'model',
        };
      }
      if (modelDriversOutsideView.length > 0) {
        return {
          state: 'ML_ONLY_OUTSIDE_VIEW',
          base: agreement,
          label: 'Model only',
          qualifier: 'on axes the statistical layer cannot see',
          reading:
            'The statistical layer does not measure ' +
            modelDriversOutsideView.join(', ') +
            ' at all, so there is nothing in the composite to corroborate or contradict this.',
          statisticalElevated: false,
          modelElevated: true,
          tone: 'model',
        };
      }
      return {
        state: 'ML_ONLY_SAME_AXIS',
        base: agreement,
        label: 'Model only',
        qualifier: 'on axes both layers share',
        reading:
          'Every axis the model isolated on is one some statistical signal also looks at, so the ' +
          'two layers are disagreeing about the same evidence rather than seeing different evidence.',
        statisticalElevated: false,
        modelElevated: true,
        tone: 'model',
      };
  }
}
